routing/bicycle_directions.cpp

```diff
@@ -222,39 +213,172 @@ Index::FeaturesLoaderGuard & BicycleDirectionsEngine::GetLoader(MwmSet::MwmId co
+      m2::PointD const & outgoingPoint = edge.GetEndJunction().GetPoint();
+      angle = my::RadToDeg(turns::PiMinusTwoVectorsAngle(junctionPoint, ingoingPoint, outgoingPoint));
+    }
+    else
+    {
+      // Note. In case of crossing mwm border
+      // (inEdge.GetFeatureId().m_mwmId != edge.GetFeatureId().m_mwmId)
+      // twins of inEdge.GetFeatureId() are considered as outgoing features.
+      // In this case that turn candidate angle is invalid and
+      // should not be used for turn generation.
+      adjacentEdges.m_outgoingTurns.isCandidatesAngleValid = false;
+    }
+    adjacentEdges.m_outgoingTurns.candidates.emplace_back(angle, uniNodeId, highwayClass);
+  }
+}
+
+Segment BicycleDirectionsEngine::GetSegment(FeatureID const & featureId, uint32_t segId, bool forward) const
+{
+  CHECK(m_numMwmIds, ());
+
+  if (!featureId.m_mwmId.IsAlive())
+    MYTHROW(RoutingException, ("Mwm:", featureId.m_mwmId, "is not alive."));
+
+  NumMwmId const numMwmId =
+    m_numMwmIds->GetId(featureId.m_mwmId.GetInfo()->GetLocalFile().GetCountryFile());
+  return Segment(numMwmId, featureId.m_index, segId, forward);
+}
+
+void BicycleDirectionsEngine::FillPathSegmentsAndAdjacentEdgesMap(
+    RoadGraphBase const & graph, vector<Junction> const & path,
+    IRoadGraph::TEdgeVector const & routeEdges, my::Cancellable const & cancellable)
+{
+  size_t const pathSize = path.size();
+  CHECK_GREATER(pathSize, 1, ());
+  CHECK_EQUAL(routeEdges.size(), pathSize - 1, ());
+  // Filling |m_adjacentEdges|.
+  m_adjacentEdges.insert(make_pair(UniNodeId(UniNodeId::Type::Mwm), AdjacentEdges(0)));
+  auto constexpr kInvalidSegId = std::numeric_limits<uint32_t>::max();
+  // |startSegId| is a value to keep start segment id of a new instance of LoadedPathSegment.
+  uint32_t startSegId = kInvalidSegId;
+  vector<Junction> prevJunctions;
+  vector<Segment> prevSegments;
+  for (size_t i = 1; i < pathSize; ++i)
+  {
+    if (cancellable.IsCancelled())
+      return;
+
+    Junction const & prevJunction = path[i - 1];
+    Junction const & currJunction = path[i];
+
+    // Note. i + 1 == pathSize means |currJunction| is a finish and the outgoing edges
+    // from finish are not important for turn generation.
+    IRoadGraph::TEdgeVector outgoingEdges;
+    if (i + 1 != pathSize)
+      graph.GetOutgoingEdges(currJunction, outgoingEdges);
+
+    IRoadGraph::TEdgeVector ingoingEdges;
+    graph.GetIngoingEdges(currJunction, ingoingEdges);
+
+    Edge const & inEdge = routeEdges[i - 1];
+    // Note. |inFeatureId| may be invalid in case of adding fake features.
+    // It happens for example near starts and a finishes.
+    FeatureID const & inFeatureId = inEdge.GetFeatureId();
+    uint32_t const inSegId = inEdge.GetSegId();
+    bool const inIsForward = inEdge.IsForward();
+
+    if (startSegId == kInvalidSegId)
+      startSegId = inSegId;
+
+    if (inFeatureId.IsValid() && i + 1 != pathSize &&
+        !IsJoint(ingoingEdges, outgoingEdges, inEdge, routeEdges[i]))
```

Comment:
> In general this is wrong, sometimes fake feature is just a half of the real segment, so if prev feature is fake, it still may be joint or not.

Yes. In this PR all points of fake features are considered as joints. So for turn generation system of each segment of fake feature is added whole instance of LoadedPathSegment.

You're right. It's possible to check if a feature is a part of real feature or not. But gathering fake features into long LoadedPathSegment takes significant efforts I guess. To do that we need (1) to convert features which part of real to real one; (2) and probably to get ready IndexRoadGraph for such cases. (I'm talking about GetOutgoingEdges() and GetIngoingEdges()).

On the other hand this extra joints should not affect to turn generation system significantly.